dts/bindings/video/video-interfaces.yaml

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+# Common properties for video interface endpoints.
+
+include: base.yaml
+
+compatible: "zephyr,video-interfaces"
```

Comment:
why have a zephyr, compatible? seems like this file should be included by actual hardware bindings, no?

Reply:
This would be for a nested node rather than the parent node, so there would still be a `compatible=` node for `vendor,device-name`:

```
&video_device {
    compatible = "vendor,device-name";
    port {
        endpoint@0 {
            compatible = "zephyr,video-interfaces";
            remote-endpoint = <&my_source>;
        };
        endpoint@1 {
            compatible = "zephyr,video-interfaces";
            remote-endpoint-label = "my_sink";
        };
    };
};
```

But `endpoint@0` and `endpoint@1` also describe properties of `vendor,device-name`.

Adding `vendor,device-name-endpoint.yaml` would then look like this?

```yaml
compatible: "vendor,device-name-endpoint"

descrption: |
   Description of what "endpoint" means in the context
  of this device (input/output? virtual channel?) 

include: "zephyr,video-interfaces.yaml"

properties:

  something-extra:
    type: int
    description: |
      As soon as the driver introduce a new property, all users would
      have to update their devicetree overlay, so having a dedicated
      compatible node helps with forward compatibility too.
    default: 7
```

I missed this completely, thank you for raising this!

Reply:
@decsny This binding is for the endpoint grandchild or child of grandchild node of the video device depending on the `ports `node is present or not (it's optional). So, if we include this file, in the bindings, we will need 2 or  3 `child-binding`" properties. The problem is we don't know 2 or 3 (depending on `ports `node present or not).

That's why I choose the alternate solution using `compabtible`

Reply:
@josuah Sorry,I can't understand your point. So, what is the problem here ?

> But endpoint@0 and endpoint@1 also describe properties of vendor,device-name
Why the child node has to describe the property of its parent ?

> Adding vendor,device-name-endpoint.yaml would then look like this?
Why does we need to add this .yaml ?

Reply:
It might not be familiar enough with devicetrees bindings... Here was my doubt:

Step 1: a driver uses `compatible = "zephyr,video-interfaces";`

```
        endpoint@0 {
            compatible = "zephyr,video-interfaces";
            data-lanes = <1 2>;
        };
```

Step 2: a driver wants extensions for configuring endpoints:

```
        endpoint@0 {
            compatible = "vendor,device-name-endpoint";
            data-lanes = <1 2>;
            buffer-size = <1024>;
        };
```

Then, would `vendor,device-name-endpoint.yaml` be able to include `zephyr,video-interfaces.yaml`? There would be 2 `compatible: "..."` then?

I probably missed something.
Thank you for your patience!

Reply:
However, a particular vendor can still use:
- `child-bindings: child-bindings:`
   to specify `port { endpoint@0 { custom = "properties"; }; };`
- `child-bindings: child-bindings: child-bindings:`
   to specify custom `ports { port@0 { endpoint@0 { custom = "properties"; }; }; };`

Like done here:
https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/onnn,ar0521.yaml#n45

Plus it matches how Linux named it: I might be learning from my mistakes after all... :)
https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/onnn,ar0521.yaml#n53

Reply:
The Linux [example](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/onnn,ar0521.yaml#n45) you give above does not add custom properties into the video-interfaces. They just use it as-is (to have bus-type and remote-endpoint). AFAIK, there are no sensor, sensor interfaces .dts in Linux that needs a custom property apart from those in video-interfaces.yaml.

I think the purpose of having a common video-interfaces binding is to standardize the interfaces properties to avoid letting sensors, sensor interfaces, ISP, et.c drivers to define their own custom properties.

As said, there are two ways to use the bindings:

- Include the entire `video-interfaces.yaml` file (so vendors will need to add 2 or 3 `child-binding` in their .dts before the include).
- Use `compatible`, we have example in the display that do this way [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/panel/panel-timing.yaml).

It's no problem to me to change to the first way (include) as both ways are same to me.

Reply:
Thank you for helping me understand Linux's logics.
If I got it right, your original proposal (1.) sounds closest to Linux and cover every use case that did show-up insofar.

Reply:
Looking at [this incomplete list](https://github.com/zephyrproject-rtos/zephyr/issues/76378):
- `clock-noncontinuous`: [sony,imx219](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/imx219.yaml) - [ovti,ov5670](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/ovti,ov5670.yaml#n87)
- `input-clock-frequency`: [aptina,mt9p031](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/aptina,mt9p031.yaml#n102)
- `pixel-clock-frequency`: [aptina,mt9p031](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/aptina,mt9p031.yaml#n102)
- `ovti,mipi-clock-voltage`: [ovti,ov02a10](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/ovti,ov02a10.yaml#n141)

Very few! And only one of them has a vendor prefix.

`ovti,ov02a10` could also add the `ovti,mipi-clock-voltage` custom properties like this:
```yaml
child-bindings:                # port { };
  child-bindings:              # port { endpoint { }; };
    properties:
      ovti,mipi-clock-voltage: # port { endpoint { ovti,mipi-clock-voltage = <>; }; };
        default: 4
```

For information, here is the list of non-vendor nodes defining child-bindings:

```
$ grep -rl child-binding dts/bindings/ | grep -v ,
```
[dts/bindings/dsa/microchip_dsa.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/dsa/microchip_dsa.yaml)
[dts/bindings/mtd/fixed-partitions.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mtd/fixed-partitions.yaml)
[dts/bindings/led/gpio-leds.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/led/gpio-leds.yaml)
[dts/bindings/led/led-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/led/led-controller.yaml)
[dts/bindings/led/pwm-leds.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/led/pwm-leds.yaml)
[dts/bindings/can/can-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/can/can-controller.yaml)
[dts/bindings/input/analog-axis.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/input/analog-axis.yaml)
[dts/bindings/input/adc-keys.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/input/adc-keys.yaml)
[dts/bindings/input/gpio-keys.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/input/gpio-keys.yaml)
[dts/bindings/adc/adc-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/adc/adc-controller.yaml)
[dts/bindings/gpio/gpio-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/gpio/gpio-controller.yaml)

Reply:
Now that you mention it it... is it impossible to have multiple compatibles? That would completely solve the problem: use the standard one as well as the extension?

https://github.com/zephyrproject-rtos/zephyr/blob/f61b8f9fd22d00246f6044182b6ef164f2fb26b3/boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts#L104

https://github.com/zephyrproject-rtos/zephyr/blob/f61b8f9fd22d00246f6044182b6ef164f2fb26b3/dts/bindings/base/zephyr%2Cmemory-region.yaml#L4-L12

https://github.com/zephyrproject-rtos/zephyr/blob/f61b8f9fd22d00246f6044182b6ef164f2fb26b3/dts/bindings/mtd/soc-nv-flash.yaml#L1-L10

Reply:
Thanks for the search ! Well, from your above examples:

> clock-noncontinuous: [sony,imx219](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/imx219.yaml) - [ovti,ov5670](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/ovti,ov5670.yaml#n87)

=> `clock-noncontinuous` is a not a custom property. It's defined [here](https://elixir.bootlin.com/linux/v6.11/source/Documentation/devicetree/bindings/media/video-interfaces.yaml#L187) in video-interfaces.yaml.

> input-clock-frequency: [aptina,mt9p031](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/aptina,mt9p031.yaml#n102)
> pixel-clock-frequency: [aptina,mt9p031](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/aptina,mt9p031.yaml#n102)

=> Looking at this sensor driver:

-  `input-clock-frequency` is the [clock](https://elixir.bootlin.com/linux/v6.11/source/drivers/media/i2c/mt9p031.c#L240) for the sensor. Don't know why they don't use the [clocks](https://elixir.bootlin.com/linux/v6.11/source/Documentation/devicetree/bindings/media/i2c/aptina,mt9p031.yaml#L94) property instead.
-  `pixel-clock-frequency` is exactly the [V4L2_CID_PIXEL_RATE] (https://elixir.bootlin.com/linux/v6.11/source/drivers/media/i2c/mt9p031.c#L1158). Normally, this is provided by the sensor driver itself to its sensor interface device to compute the frame rate (see https://elixir.bootlin.com/linux/v6.10-rc1/source/drivers/media/i2c/ov5640.c#L3448, for example). It's not something the sensor needs to read from devicetree.

> ovti,mipi-clock-voltage: [ovti,ov02a10](https://git.kernel.org//pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/media/i2c/ovti,ov02a10.yaml#n141)

The only place I see it really needs a new custom property is in this sensor. But I suppose it can be done differently as well.

So, I keep my thinking that "the purpose of having a common video-interfaces binding is to standardize the interfaces properties to avoid letting sensors, sensor interfaces, ISP, et.c drivers to define their own custom properties." :).

But well, it is safe and does not has any impact if we choose other way (include the .yaml). I will change to this.

Reply:
> [the] purpose of having a common video-interfaces binding is to standardize the endpoint interface properties to avoid letting sensors, sensor interfaces, ISP, etc. drivers to define their own custom properties
I was curious about possible consequences, and I am glad to see such investigation.

@decsny given the last message of @ngphibang, does it look reasonable to omit a per-sensor endpoint bindings, and instead rely on `child-bindings:` to [inject custom properties from the parent node](https://github.com/zephyrproject-rtos/zephyr/pull/74415#discussion_r1771436310)?

Reply:
@decsny : I tried to include the file in the a video device .yaml, for example, in `ovti,ov5640.yaml`., instead of using `compatible ` as your request:

```
child-binding:
  child-binding:
      include: video-interfaces.yaml
```

but I  got error:

`devicetree error: 'compatible' is marked as required in 'properties:' in C:/Users/nxf91161/zephyrproject/zephyr/dts/bindings\video\ovti,ov5640.yaml, but does not appear in <Node /soc/i2c@40c38000/ov5640@3c/port/endpoint`

It complains because the `ovti,ov5640.yaml` requires a `compatible `property in its endpoint child node so it comes back to the "compatible" way.

And as discussed above, we see no advantages or drawbacks between "**include**" and "**compatible**" ways. Could you give the reason why we should "include" and not use "compatible" ? Given that we have an example in the display that do "`compatible`" way [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/panel/panel-timing.yaml).

Reply:
@ngphibang can you show the DT you have for the ov5460 node and child nodes where you got that error, I don't understand why you got that error, the ov5460 binding should not be being applied to the endpoint node?

Reply:
@decsny : Here it is (nxp_btb44_ov5640.overlay):

```
&nxp_cam_i2c {
	status = "okay";

	ov5640: ov5640@3c {
		compatible = "ovti,ov5640";
		reg = <0x3c>;
		reset-gpios = <&nxp_cam_connector 9 GPIO_ACTIVE_LOW>;
		powerdown-gpios = <&nxp_cam_connector 17 GPIO_ACTIVE_HIGH>;

		port {
			ov5640_ep_out: endpoint {
				remote-endpoint-label = "mipi_csi2rx_ep_in";
				bus-type = "mipi-csi2-dphy";
				data-lanes = <1 2>;
			};
		};
	};
};

&nxp_mipi_csi {
	status = "okay";

	sensor = <&ov5640>;

	ports {
		port@1 {
			reg = <1>;

			mipi_csi2rx_ep_in: endpoint {
				remote-endpoint-label = "ov5640_ep_out";
				data-lanes = <1 2>;
			};
		};
	};
};
```

And here is the entire ovti,ov5640.yaml

```
description: OV5640 CMOS video sensor

compatible: "ovti,ov5640"

include: i2c-device.yaml

properties:
  reset-gpios:
    type: phandle-array
    description: |
      The RESETB pin is asserted to cause a hard reset. The sensor
      receives this as an active-low signal.
  powerdown-gpios:
    type: phandle-array
    description: |
      The PWDN pin is asserted to disable the sensor. The sensor
      receives this as an active-high signal.

child-binding:
  child-binding:
    include: video-interfaces.yaml
```

Though it does not relate to this PR as you said, it's interesting to know why it requires the child node having a "compatible" properties. I think it come from "include: i2c-device.yaml" which in turn, include base.yaml who requires compatible. I also tried to move the "child-binding" block before "include: i2c-device.yaml" but it's the same

Reply:
it's simply just because zephyr's DT binding schemas are not as powerful as linux, in linux for example you can make a property a node, which is what you would want to do in this case, instead we have to use a compatible.